File: androidApp/src/main/kotlin/se/soderbjorn/lunamux/android/net/NetworkStatus.kt
/**
 * Coarse connectivity classification for connection-failure messaging.
 *
 * When a connect attempt fails, the hosts screen uses these helpers to tell
 * the user *why* it probably failed ("you're on mobile data") instead of
 * showing a generic timeout. Requires only the normal-level
 * `ACCESS_NETWORK_STATE` permission — no runtime prompt.
 *
 * @see se.soderbjorn.lunamux.android.ui.HostsScreen
 */
package se.soderbjorn.lunamux.android.net

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities

/**
 * Snapshot queries over the active network's transport. Each call reads the
 * live [ConnectivityManager] state; failures (missing service, stale
 * network handle) degrade to `false` so callers fall back to generic copy.
 */
object NetworkStatus {

    /**
     * Whether the phone's active network rides Wi-Fi.
     *
     * @param context any context; the application context is used internally.
     * @return `true` when the active network has the Wi-Fi transport.
     */
    fun isOnWifi(context: Context): Boolean =
        hasTransport(context, NetworkCapabilities.TRANSPORT_WIFI)

    /**
     * Whether the phone's active network rides mobile data.
     *
     * @param context any context; the application context is used internally.
     * @return `true` when the active network has the cellular transport.
     */
    fun isOnCellular(context: Context): Boolean =
        hasTransport(context, NetworkCapabilities.TRANSPORT_CELLULAR)

    /** Shared active-network transport probe backing the public queries. */
    private fun hasTransport(context: Context, transport: Int): Boolean = runCatching {
        val cm = context.applicationContext
            .getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            ?: return false
        val network = cm.activeNetwork ?: return false
        cm.getNetworkCapabilities(network)?.hasTransport(transport) == true
    }.getOrDefault(false)
}

File: iosApp/iosApp/Net/NetworkStatus.swift
/// Coarse connectivity classification for connection-failure messaging.
///
/// When a connect attempt fails, the hosts screen uses this to tell the user
/// *why* it probably failed ("you're on mobile data") instead of surfacing a
/// generic timeout. Mirrors the Android `NetworkStatus` object, which reads
/// `ConnectivityManager` for the same purpose.

import Foundation
import Network

/// Live view of the current network path's transport.
///
/// Unlike Android's `ConnectivityManager`, `NWPathMonitor` is push-based: it
/// has no synchronous "what am I on right now" query, so a single long-lived
/// monitor keeps the last known path cached for the failure path to read. The
/// monitor is started once at first use and runs for the process lifetime —
/// it is cheap, and the alternative (spinning one up inside the error handler)
/// would race the very answer we need.
///
/// Failures degrade to `false` so callers fall back to generic copy.
final class NetworkStatus: @unchecked Sendable {
    static let shared = NetworkStatus()

    private let monitor = NWPathMonitor()
    private let lock = NSLock()
    private var currentPath: NWPath?

    private init() {
        monitor.pathUpdateHandler = { [weak self] path in
            guard let self else { return }
            lock.lock()
            currentPath = path
            lock.unlock()
        }
        monitor.start(queue: DispatchQueue(label: "se.soderbjorn.lunamux.networkstatus"))
    }

    /// Whether the device's active path rides Wi-Fi.
    var isOnWifi: Bool { usesInterface(.wifi) }

    /// Whether the device's active path rides mobile data.
    var isOnCellular: Bool { usesInterface(.cellular) }

    /// Shared active-path transport probe backing the public queries.
    ///
    /// - Parameter type: the interface type to test for.
    /// - Returns: `true` when the satisfied path uses that interface.
    private func usesInterface(_ type: NWInterface.InterfaceType) -> Bool {
        lock.lock()
        let path = currentPath
        lock.unlock()
        guard let path, path.status == .satisfied else { return false }
        return path.usesInterfaceType(type)
    }
}
